service/tablet_allocator.cc

```diff
@@ -825,9 +825,8 @@ class tablet_allocator_impl : public tablet_allocator::impl
     }
 
     void on_before_create_column_family(const keyspace_metadata& ksm, const schema& s, std::vector<mutation>& muts, api::timestamp_type ts) override {
-        keyspace& ks = _db.find_keyspace(s.ks_name());
-        auto&& rs = ks.get_replication_strategy();
-        if (auto&& tablet_rs = rs.maybe_as_tablet_aware()) {
+        auto rs = abstract_replication_strategy::create_replication_strategy(ksm.strategy_name(), ksm.strategy_options());
```

Comment:
@tgrabiec please review this patch (` tablet_allocator: update on_before_create_column_family `)

Reply:
Is this safe to do even in the case where keyspace already exists?

Reply:
If it is unsafe, can we just check whether the keyspace exists and run the old code if it does? If we can, then the comment above `migration_listener::on_before_create_column_family`:
```
// ... Therefore, listeners can't load the keyspace from the database.
// Instead, they should use the `ksm` parameter if needed.
```
is misleading and I should change it.

Reply:
> If it is unsafe

I don't know that. You should investigate it.

The question is, in the case of keyspace already existing, whether this `create_replication_strategy` will return the same thing as `ks.get_replication_strategy()` would. If it turns out these replication strategies are different in a way that may affect the behavior of below code, this is unsafe.

Even if the keyspace does not exist yet, it could be that creating it first and then calling `get_replication_strategy()` would return a different thing than doing `create_replication_strategy` directly here, again causing different behavior of below code. (I don't believe that this happens though, after all, whatever `get_replication_strategy()` returns must have been created somehow, we just need to find a way to recreate the same thing. But you should check.)

Reply:
> The question is, in the case of keyspace already existing, whether this `create_replication_strategy` will return the same thing as `ks.get_replication_strategy()` would. If it turns out these replication strategies are different in a way that may affect the behavior of below code, this is unsafe.

Let's assume that the keyspace exists.

If the new `prepare_new_column_family_announcement` is called with `keyspace_metadata` different than the actual metadata, then of course these strategies can be different. I don't know why someone would do this though.

If the new `prepare_new_column_family_announcement` is called with actual `keyspace_metadata`, I'm not sure if these strategies can be different. The question is whether a keyspace's replication strategy always stays consistent with its metadata. In other words, whether `ks.get_replication_strategy()` and `abstract_replication_strategy::create_replication_strategy(ks.metadata()->strategy_name(), ks.metadata()->strategy_options())` always describe identical strategies. I think it's true, but unfortunately I couldn't find a proof.

If the above is true, we can solve the first case by prohibiting a call to the new `prepare_new_column_family_announcement` if the keyspace exists (for example, by throwing an exception).

Otherwise, we cannot use the `keyspace_metadata` parameter in `tablet_allocator_impl::on_before_create_column_family` if the keyspace exists even if it came from the old `prepare_new_column_family_announcement`. The only solution I see is using the `keyspace_metadata` parameter there only if the keyspace doesn't exist. I like this solution less because it is less general and doesn't prevent similar problems from appearing in other places.

> Even if the keyspace does not exist yet, it could be that creating it first and then calling `get_replication_strategy()` would return a different thing than doing `create_replication_strategy` directly here, again causing different behavior of below code. (I don't believe that this happens though, after all, whatever `get_replication_strategy()` returns must have been created somehow, we just need to find a way to recreate the same thing. But you should check.)

Let's assume that the keyspace doesn't exist.

`prepare_new_keyspace_announcement` calls `make_create_keyspace_mutations` that creates mutations for the strategy options and the strategy name using `keyspace_metadata` in exactly the opposite way to how `create_keyspace_from_schema_partition` creates `keyspace_metadata`. In `database::parse_system_tables`, `keyspace_metadata` received from `create_keyspace_from_schema_partition` is passed unchanged to `create_keyspace` and to `create_in_memory_keyspace` that calls `keyspace::create_replication_strategy`, which uses the strategy name and the strategy options from the unchanged metadata to create the replication strategy.

So, it looks like calling `get_replication_strategy()` after creating the keyspace and calling `create_replication_strategy` directly would return the same strategy.

Reply:
After discussing it with @kbr-scylla, we concluded that:
- we can safely use `keyspace_metadata` and the `abstract_replication_strategy::create_replication_strategy` parameter in `tablet_allocator_impl::on_before_create_column_family` assuming the keyspace doesn't exist or it exists and `keyspace_metadata` is actual,
- in `tablet_allocator_impl::on_before_create_column_family`, calling `ks.get_replication_strategy()` after creating the keyspace and calling `abstract_replication_strategy::create_replication_strategy` directly when the keyspace doesn't exist returns the same strategy.

So, we only have to worry about this:
> Let's assume that the keyspace exists.
>
> If the new `prepare_new_column_family_announcement` is called with `keyspace_metadata` different than the actual metadata, then of course these strategies can be different. I don't know why someone would do this though.

We decided that a simple sanity check, that ensures the new `prepare_new_column_family_announcement` is called only when the keyspace doesn't exist, is enough. After all, this function has been introduced only to handle a case when the keyspace doesn't exist.